spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/util/MaterializingIterator.scala

```diff
@@ -0,0 +1,37 @@
+package com.datastax.spark.connector.util
+
+import java.util.NoSuchElementException
+
+import scala.collection.mutable
+
+/** Materializes/Fetches up to `materializationLimit` from underlying `iterator`.
+  * Materialization happens on `next()` invocation.
+  *
+  * In following example `it.next()` is invoked three times, results are caches in internal queue.
+  * {{{
+  * new MaterializingIterator(it, 3).next()
+  * }}}
+  *
+  */
+class MaterializingIterator[T](iterator: Iterator[T], materializationLimit: Int) extends Iterator[T] {
+
+  assert(materializationLimit > 0, "Materialization limit must be greater than 0")
+
+  private val materialized: mutable.Queue[T] = new mutable.Queue[T]()
+
+  private def materialize(): Unit = {
+    while (iterator.hasNext && materialized.size < materializationLimit) {
+      materialized.enqueue(iterator.next)
+    }
+  }
+
+  override def hasNext = !materialized.isEmpty || iterator.hasNext
+
+  override def next() = {
+    if(!hasNext) {
+      throw new NoSuchElementException("No more elements")
+    }
+    materialize()
```

Comment:
@maasg low watermark looks good, if this way of fetching rows holds I'll add it. 

@RussellSpitzer the intention of this iterator was to avoid memory overfilling by keeping number of created/fetched objects under control. With async queries for all `left` elements, aren't we at risk of OutOfMemory errors?
WDYT about modifying this by adding separate thread that would be responsible for filling up queue?